File: dp/day1_fib_number/509.fibonacci-number.py
#
# @lc app=leetcode id=509 lang=python3
#
# [509] Fibonacci Number
#

# @lc code=start
from functools import cache


class Solution:
    @cache
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        return self.fib(n-1)+self.fib(n-2)
        
# @lc code=end

File: dp/day1_fib_number/509.fibonacci-number.cpp
/*
 * @lc app=leetcode id=509 lang=cpp
 *
 * [509] Fibonacci Number
 */

#include "bits/stdc++.h"

using namespace std;

// @lc code=start
class Solution {
public:
  int fib(int n) {
    vector<int> dp(n + 1, -1);
    return fibTopDown(dp, n);
  }

  int fibTopDown(vector<int> &dp, int n) {
    if (n <= 1) {
      return n;
    }

    if (dp[n] == -1) {
      dp[n] = fibTopDown(dp, n - 2) + fibTopDown(dp, n - 1);
    }

    return dp[n];
  }
};
// @lc code=end
